floyd/cli/experiment.py

```diff
@@ -172,26 +162,47 @@ def logs(id, url, tail, follow, sleep_duration=1):
 
         sleep(1)
 
-    log_url = "{}/api/v1/resources/{}?content=true".format(
-        floyd.floyd_host, instance_log_id)
+    return instance_log_id
+
+
+def follow_logs(instance_log_id, sleep_duration=1):
+    cur_idx = 0
+    while True:
+        # Get the logs in a loop and log the new lines
+        log_file_contents = ResourceClient().get_content(instance_log_id)
+        print_output = log_file_contents[cur_idx:]
+        cur_idx += len(print_output)
+        if len(print_output.strip()):
+            floyd_logger.info(print_output.rstrip())
+        sleep(sleep_duration)
+
+
+@click.command()
+@click.option('-u', '--url', is_flag=True, default=False, help='Only print url for accessing logs')
+@click.option('-t', '--tail', is_flag=True, default=False, help='Stream the logs')
+@click.option('-f', '--follow', is_flag=True, default=False, help='Stream the logs (alias for -t/--tail)')
+@click.argument('id', nargs=1, required=False)
+def logs(id, url, tail, follow, sleep_duration=1):
+    """
+    Print the logs of the run.
+    """
+    tail = tail or follow
+
+    instance_log_id = get_log_id(id)
+
     if url:
+        log_url = "{}/api/v1/resources/{}?content=true".format(
+            floyd.floyd_host, instance_log_id)
         floyd_logger.info(log_url)
         return
+
     if tail:
         floyd_logger.info("Launching job ...")
-        current_shell_output = ""
-        while True:
-            # Get the logs in a loop and log the new lines
-            log_file_contents = ResourceClient().get_content(instance_log_id)
-            print_output = log_file_contents[len(current_shell_output):]
-            if len(print_output.strip()):
-                floyd_logger.info(print_output)
-            current_shell_output = log_file_contents
-            sleep(sleep_duration)
+        follow_logs(instance_log_id, sleep_duration)
     else:
         log_file_contents = ResourceClient().get_content(instance_log_id)
         if len(log_file_contents.strip()):
-            floyd_logger.info(log_file_contents)
+            floyd_logger.info(log_file_contents.rstrip())
```

Comment:
What's the reason for adding a rstrip here?

Reply:
nvm, just saw your comment. if the newline unexpected? I feel like if the original log file has new lines, we should display it as is. In terms of showing content generated by user, format prettifying should not be applied by default. If the original log file doesn't contain new lines at the end, then it's something I can help fix in the backend :)

Reply:
From what I can tell, `ResourceClient().get_content` always returns lines, right? The last of those lines probably will have a newline at the end, but then `floyd_logger.info` will add a newline of its own, resulting in one too many. The problem with this currently can be seen with e.g.:
```
floyd run 'bash -c "while true; do date; sleep 1; done"'
floyd logs follow <id>
...
2018-02-15 22:02:28,390 INFO - Run Output:
2018-02-15 22:02:28,394 INFO - Fri Feb 16 06:02:28 UTC 2018
2018-02-15 22:02:29,394 INFO - Fri Feb 16 06:02:29 UTC 2018
2018-02-15 22:02:30,395 INFO - Fri Feb 16 06:02:30 UTC 2018
2018-02-15 22:02:31,398 INFO - Fri Feb 16 06:02:31 UTC 2018
2018-02-15 22:02:32,399 INFO - Fri Feb 16 06:02:32 UTC 2018
2018-02-15 22:02:33,403 INFO - Fri Feb 16 06:02:33 UTC 2018
2018-02-15 22:02:34,404 INFO - Fri Feb 16 06:02:34 UTC 2018

2018-02-15 22:02:35,406 INFO - Fri Feb 16 06:02:35 UTC 2018
2018-02-15 22:02:36,408 INFO - Fri Feb 16 06:02:36 UTC 2018

2018-02-15 22:02:37,411 INFO - Fri Feb 16 06:02:37 UTC 2018
2018-02-15 22:02:38,413 INFO - Fri Feb 16 06:02:38 UTC 2018

2018-02-15 22:02:39,415 INFO - Fri Feb 16 06:02:39 UTC 2018
2018-02-15 22:02:40,417 INFO - Fri Feb 16 06:02:40 UTC 2018
```
And I'm guessing that `get_content` always returns text ending in a newline because somewhere the newline is the sign to flush the output buffer?

Reply:
Thanks for catching this, this seems like a bug from the backend, let me dig into. Will keep you posted on the update.

Reply:
Hey, any update on this? It would be cool to get this finalised (though I know you guys probably have a lot to be doing at the moment - thanks for all the hard work!)

Reply:
Hey, sorry for the delay. Turns out that new line is expected, because date command also prints out a new line at the end. I have sent a PR to your branch to further simplify the code so we can show the raw log content to the terminal. Let me know if my PR makes sense to you :)

Reply:
We were busy launching realtime support for our web UI last week, so I got distracted. Hope you like the new UX.

Reply:
No worries :) The new logs box is niiice.

Pull request looks good. Let's do it.